ch01python/00pythons.ipynb.py

```diff
@@ -12,224 +12,237 @@
+    return fig
 
 
 # %% [markdown]
-# In a real example, we could edit the file on disk
-# using a code editor such as [VS code](https://code.visualstudio.com/).
+# We can use this code in the notebook by _importing_ the `draw_eight` module and then _calling_ the
+# `make_figure` function defined in the module.
 
 # %%
-import draw_eight # Load the library file we just wrote to disk
-
-# %%
-image = draw_eight.make_figure()
+import draw_eight  # Load the library
+fig = draw_eight.make_figure()
 
 # %% [markdown]
-# Note, we can import our `draw_eight` module in this notebook only if the file is in our current working directory (i.e. the folder this notebook is in).
-#
-# To allow us to import our module from anywhere on our computer, or to allow other people to reuse it on their own computer, we can create a [Python package](https://packaging.python.org/en/latest/).
+# We will cover how to import and use functionality from libraries, how to install third-party libraries
+# and how to write your own libraries that can be shared and used by other in this course.
 #
-
-# %% [markdown]
-# ### Python packages
-#
-# A package is a collection of modules that can be installed on our computer and easily shared with others. We will learn how to create packages later on in this course.
-#
-# There is a huge variety of available packages to do pretty much anything. For instance, try `import antigravity` or `import this`.
+# ### Python scripts
 #
+# While Jupyter notebooks are a great medium for learning how to use Python and for exploratory work,
+# there are some drawbacks:
+#
+#   * The require Jupyter Lab (or a similar application) to be installed to run the notebook.
+#   * It can be difficult to run notebooks non-interactively, for example when scheduling a job on a cluster [such as those offered by UCL Research Computing](https://www.rc.ucl.ac.uk/docs/Background/Cluster_Computing).
+#   * The flexibility of being able to run the code in cells in any order can also make it difficult to reason how outputs were produced and can lead to non-reproducible analyses.
+#   
+# In some settings it can therefore be preferrable to write Python _scripts_ - that is files (typically with
+# a `.py` extension) which contain Python code which completely describes a computational task to perform
+# and that can be run by  passing the name of the script file to the `python` program in a command-line
+# environment. Optionally scripts may also allow passing in arguments from the command-line to control
+# the execution of the script. As scripts are generally run from text-based terminals, non-text outputs such
+# as images will generally be saved to files on disk.
+#
+# Python scripts are well suited to for example for describing computationally demanding simulations or analyses
+# to run as long jobs on a remote server or cluster, or tasks where the input and output is mainly at the file level
+# - for instance batch processing a series of data files.
```

Comment:
```suggestion
# - for instance batch processing a series of data files.
#
# ### IDEs
# 
# IDEs are Interactive Development Environments and it's what we will be using in this course.
# We will be demonstrating it through [VS Code](https://code.visualstudio.com/) but you could use whichever you like, e.g., [spyder](https://www.spyder-ide.org/), [pycharm](https://www.jetbrains.com/pycharm/), ...).
# We won't be using notebooks, except for this notes so you can download and experiment with them.
# However, we will be learning how to build libraries, and they need to be composed of python files rather than notebooks.
# When working with an IDE, you'll get access to a Python interpreter and you can run scripts directly from the interface as well as use tools like the debugger, test frameworks and git from within it.
```